sui-execution/src/tests.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeSet, HashMap},
+    path::PathBuf,
+};
+
+use cargo_metadata::{DependencyKind, Metadata, MetadataCommand, Package};
+use petgraph::{algo::all_simple_paths, prelude::DiGraphMap};
+
+type PackageGraph<'p> = DiGraphMap<&'p str, ()>;
+
+struct Packages(HashMap<String, Package>);
+
+#[test]
+/// Make sure that all accesses to execution layer crates in the `sui-node` and `sui-replay` crates
+/// go via the `sui-execution` crate (in other words, the `sui-execution` crate dominates execution
+/// layer crates in the dependency graphs of `sui-node` and `sui-replay`).
+///
+/// This helps ensures that execution that may be committed on-chain respects the execution version
+/// that is stated in the protocol config.
+fn test_encapsulation() {
```

Comment:
This is the main motivating change for this PR -- the rest of the changes are updates to dependencies driven by this test.

@bmwill, I looked at using `guppy` for this initially, but I found that I needed to carve out sub-graphs and run algorithms on those.  `petgraph` was ideal for that, and using `cargo_metadata` seemed ideal for populating `petgraph`'s graphs, but let me know if I missed something.

Reply:
Guppy uses both cargo-metadata and petgraph internally to allow for you to do this much simpler imo as it provides convenience tools for these things. Also i think this would be better suited as a guppy lint (see the `x` crate for this) as it'll get run much earlier in the testing pipeline

Reply:
Guppy also handles a number of very awkward edges cases for dealing with dependency resolution

Reply:
If you can give me a pointer to how to do this with `guppy`, I'm happy to give it a try.  The query is quite simple: The crates that `sui-execution` depends on (minus a fixed set) should not be accessible in the package graphs of `sui-node` and `sui-replay` if `sui-execution` has been removed from them.

Alternatively/equivalently, the `sui-execution` crate needs to dominate all the execution crates in the package graphs of `sui-node` and `sui-replay`.

I couldn't find an easy way of encoding the "remove `sui-execution` from the package graph" or the "dominates" properties in `guppy`, so that's why I reached for `cargo_metadata` and `petgraph`.

Reply:
I gave `guppy` a second try, I got a lot closer thanks to your help, but unfortunately, after I added path printing it became clear that it was reporting false-positives by following `build-dependencies`.

When I tried to address that, I ended up with the opposite problem: The test seemed to identify an issue but couldn't recover the path.  Given that I don't think we'll need the richer package information that `guppy` provides here, and the guppy version is starting to look more complex than what I originally wrote (granted this is subjective), I think I'm going to stick with what we've got here.

Here's where I got with the `guppy` impl though:

```rust
// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use std::collections::{HashMap, HashSet};

use guppy::graph::{DependencyDirection, PackageLink};
use nexlint::prelude::{
    LintFormatter, LintLevel, Linter, ProjectContext, ProjectLinter, Result, RunStatus,
};

#[derive(Debug)]
/// Make sure that all accesses to execution layer crates in the `sui-node` and `sui-replay` crates
/// go via the `sui-execution` crate (in other words, the `sui-execution` crate dominates execution
/// layer crates in the dependency graphs of `sui-node` and `sui-replay`).
///
/// This helps ensures that execution that may be committed on-chain respects the execution version
/// that is stated in the protocol config.
pub(crate) struct SuiExecutionLint;

impl Linter for SuiExecutionLint {
    fn name(&self) -> &'static str {
        "sui-execution-lint"
    }
}

impl ProjectLinter for SuiExecutionLint {
    fn run<'l>(
        &self,
        ctx: &ProjectContext<'l>,
        out: &mut LintFormatter<'l, '_>,
    ) -> Result<RunStatus<'l>> {
        use DependencyDirection as DD;
        let allowed_packages: HashSet<_> = [
            "sui-protocol-config",
            "sui-types",
            "move-binary-format",
            "move-bytecode-utils",
            "move-core-types",
            "move-vm-config",
            "workspace-hack",
        ]
        .into_iter()
        .collect();

        let package_graph = ctx.package_graph()?;
        let workspace_hack = ctx.workspace_hack_name();

        let execution_crates = package_graph
            .query_workspace_names(["sui-execution"])
            .unwrap()
            .resolve_with_fn(|_, link| {
                if link.dev_only() {
                    return false;
                }

                // Collect direct dependencies of workspace packages.
                let (from, to) = link.endpoints();

                if from.name() != "sui-execution" {
                    return false;
                }

                // Skip the workspace hack package if it exists
                if let Some(workspace_hack_name) = workspace_hack {
                    if to.name() == workspace_hack_name {
                        return false;
                    }
                }

                if !to.source().is_local() {
                    return false;
                }

                !allowed_packages.contains(to.name())
            });

        fn is_bypass_execution_link(workspace_hack: Option<&str>, link: &PackageLink) -> bool {
            if link.normal().status().is_never() {
                return false;
            }

            // Collect direct dependencies of workspace packages.
            let (_from, to) = link.endpoints();

            // Skip the workspace hack package if it exists
            if let Some(workspace_hack_name) = workspace_hack {
                if to.name() == workspace_hack_name {
                    return false;
                }
            }

            if !to.source().is_local() {
                return false;
            }

            to.name() != "sui-execution"
        }

        // get set of packages excluding sui-execution
        let bypass_sui_execution = package_graph
            .query_workspace_names(["sui-node", "sui-replay"])
            .unwrap()
            .resolve_with_fn(|_, link| is_bypass_execution_link(workspace_hack, &link));

        let problem_execution_crates = bypass_sui_execution.intersection(&execution_crates);

        if problem_execution_crates.is_empty() {
            // No issues, finish.
            return Ok(RunStatus::Executed);
        }

        // Build up examples of problem paths in the package graph, from the bottom up.
        let mut suffixes: HashMap<&str, HashSet<Vec<&str>>> = problem_execution_crates
            .packages(DD::Forward)
            .map(|p| (p.name(), HashSet::from_iter([vec![p.name()]])))
            .collect();

        // Traverse links in reverse topological order
        for link in bypass_sui_execution.links(DD::Reverse) {
            if !is_bypass_execution_link(workspace_hack, &link) {
                continue;
            }

            let (from, to) = link.endpoints();
            let Some(paths) = suffixes.get(to.name()).cloned() else {
                continue;
            };

            suffixes
                .entry(from.name())
                .or_default()
                .extend(paths.into_iter().map(|mut p| {
                    p.push(from.name());
                    p
                }));
        }

        out.write(
            LintLevel::Error,
            "protocol-sensitive binaries depend on execution crates outside of 'sui-execution':",
        );

        for root in bypass_sui_execution.root_packages(DD::Forward) {
            for path in suffixes.get(root.name()).into_iter().flatten() {
                let mut path = path.clone();
                path.reverse();
                out.write(LintLevel::Error, path.join(" -> "))
            }
        }

        out.write(
            LintLevel::Error,
            "This can cause execution to fork by not especting the execution layer version set in \
             the protocol config.  Fix this by depending on these crates via 'sui-execution'.",
        );

        out.write(
            LintLevel::Error,
            "P.S. if you believe one of these crates should not be part of 'sui-execution' then \
             update the lint to exclude this crate.",
        );

        Ok(RunStatus::Executed)
    }
}
```